File: Rust/src/info/tableinfo.rs
use sqlx::sqlite::SqliteConnection;
use sqlx::{ Error, Cursor, row::Row };
use std::collections::HashMap;

pub struct TableInfo {
    name: String,
    num_rows: i32,
    columns: HashMap<String, String>
}

impl TableInfo {
    async fn read_rows(&mut self, connection: &mut SqliteConnection) -> Result<(), Error> {
        let sql = format!(r"SELECT COUNT(*) FROM '{}'", &self.name);
        let mut rows = sqlx::query(&sql).fetch(connection);

        if let Some(row) = rows.next().await? {
            self.num_rows = row.get(0);
        }                        
        Ok(())
    }
    async fn read_columns(&mut self, connection: &mut SqliteConnection) -> Result<(), Error> {
        let sql = format!(r"PRAGMA table_info('{}')", &self.name);
        let mut rows = sqlx::query(&sql).fetch(connection);

        while let Some(row) = rows.next().await? {
            self.columns.insert(row.get(1), row.get(2));
        }                        
        Ok(())
    }
    pub async fn load_from_db(&mut self, table_name: &str, connection: &mut SqliteConnection) -> Result<(), Error> {
        self.name = String::from(table_name);
        self.read_rows(connection).await?;               
        self.read_columns(connection).await?;               
        Ok(())
    }
    pub fn compare_tables(&self, table_to_check: &TableInfo) -> bool {
        if self.num_rows != table_to_check.num_rows {
            return false;
        }
        if self.columns.len() == table_to_check.columns.len() {
            for (col_name, col_type) in &self.columns {
                let col_type2 = table_to_check.columns.get(col_name);
                if col_type2.is_none() || (col_type2 != Some(col_type)) {
                    return false;
                }
            }
        } else {
            return false;
        }
        true
    }
    pub fn display_str(&self, include_name: bool) -> String {
        if include_name {
            return format!("{} has {} rows and {} columns", self.name, self.num_rows, self.columns.len());
        } else {
            return format!("has {} rows and {} columns", self.num_rows, self.columns.len());
        }
    }
    pub fn get_num_rows(&self) -> i32 { self.num_rows }
    pub fn get_num_columns(&self) -> usize { self.columns.len() }
}

impl Default for TableInfo {
    fn default() -> Self {
        TableInfo {
            num_rows: 0,
            name: String::from(""),
            columns: HashMap::new()
        }
    }
}

File: Java/src/main/java/dbComp/TableInfo.java
package dbComp;

import java.sql.Connection;
import java.sql.ResultSet;
import java.sql.SQLException;
import java.sql.Statement;
import java.util.AbstractMap;
import java.util.Iterator;
import java.util.List;
import java.util.Vector;


class TableInfo
{
    String name;
    int numRows = 0;
    List<AbstractMap.SimpleEntry<String, String>> columns = new Vector<AbstractMap.SimpleEntry<String, String>>();

    public TableInfo(Connection sqlconn, String name) throws SQLException
    {
        this.name = name;
        numRows = getNumRows(sqlconn);
        Statement stmt  = sqlconn.createStatement();
        String sql = "PRAGMA table_info('" + this.name + "')";
        ResultSet rs = stmt.executeQuery(sql);

        while (rs.next())
        {
            columns.add(new AbstractMap.SimpleEntry<String, String>(rs.getString(1), rs.getString(2)));
        }
    }
    public int numColumns() {
        return columns.size();
    }
    int getNumRows(Connection sqlconn) throws SQLException
    {
        Statement stmt  = sqlconn.createStatement();
        String sql = "SELECT COUNT(*) FROM '" + this.name + "'";
        ResultSet rs = stmt.executeQuery(sql);
        if (rs.next())
            return rs.getInt(1);
        return 0;
    }
    public boolean compare(TableInfo table)
    {
        if ((numRows == table.numRows) && (columns.size() == table.columns.size()))
        {
            Iterator<AbstractMap.SimpleEntry<String, String>> it = table.columns.iterator();
            for (AbstractMap.SimpleEntry<String, String> colInfo : columns)
            {
                if (!colInfo.equals(it.next()))
                    return false;
            }
            return true;
        }
        else
            return false;
    }
    public String getDescription()
    {
        return String.format("has %d rows and %d columns", numRows, columns.size());
    }
}
